### DISCUSSION TITLE
Usage as Memory + Disk Cache

### DISCUSSION THREAD

**Discussion by seyfeb** [ZiggyCreatures/FusionCache#72] [Q&A] [answered]
We are considering using FusionCache in a mobile application with the `IMemoryCache` and a file-based 2nd-level cache. If I understood correctly, to this end I would need to supply an implementation of `IDistributedCache` which actually caches the data on disk (e.g., via SQLite).

Is there any issue with this usage I’m not / I should be aware of?

Is there a way to automatically or manually cleanup / delete all data which has expired more than some time span to keep the disk pressure to a minimum?

**Comment by jodydonetti** [comment, marked as answer]
Hi @seyfeb and thanks for considering FusionCache!

> We are considering using FusionCache in a mobile application with the IMemoryCache and a file-based 2nd-level cache. If I understood correctly, to this end I would need to supply an implementation of IDistributedCache which actually caches the data on disk (e.g., via SQLite).

Exactly! I collected the available `IDistributedCache` implementations - at least the ones I've been able to find - in the related [docs page](https://github.com/jodydonetti/ZiggyCreatures.FusionCache/blob/main/docs/CacheLevels.md).
Here you can find the [NeoSmart.Caching.Sqlite](https://www.nuget.org/packages/NeoSmart.Caching.Sqlite/) package: there's an [official repo](https://github.com/neosmart/AspSqliteCache) on GH (updated as recently as April of this year) and an introductory [blogpost](https://neosmart.net/blog/2019/sqlite-cache-for-asp-net-core/).

I've never worked with it, but I'll take a look at their code to see if I see something smelly and come back to you.

> Is there any issue with this usage I’m not / I should be aware of?

Nope, that is the right approach 👍

> Is there a way to automatically or manually cleanup / delete all data which has expired more than some time span to keep the disk pressure to a minimum?

I'm not sure I understood what you mean.

If you mean directly with FusionCache then yes, just specify the `Duration` and the data will expire and get cleaned up automatically for you (provided the specific `IDistributedCache` impl works correctly, and I will take a look at it asap).

If instead you mean directly on SQLite it depends on the implementation and the specific commands available.

Hope this helps, will get back to you asap after I took a look at the package I mentioned.


**NOTE:** I'll add that as of a couple of weeks ago you can even specify a different `DistributedCacheDuration` so that you can have, for example, memory cache expire sooner and distributed cache expire later on, but I don't think this is what you'll need in a mobile since it's something more useful in multi-node deployments where the distributed cache can be updated by other nodes and you want to keep data in-sync without using, for whatever reason, a [backplane](https://github.com/jodydonetti/ZiggyCreatures.FusionCache/blob/main/docs/Backplane.md).

**Comment by jodydonetti** [comment]
I'm quickly looking at it right now, and in the `SqliteCacheOptions` object you can specify how frequently it'll actually remove the data that is expired, via ([`CleanupInterval`](https://github.com/neosmart/AspSqliteCache/blob/master/SqliteCache/SqliteCacheOptions.cs#L43)) (default `30 min`).

So by default every 30 min it will physically remove the data from the db, but if you care about keeping the filesize as small as possible you may drop that value to something like `5 min` or even less. The cleanup operation (that runs in the background, so will not block your code) may take some cpu cycles, but if done frequently there should be not that much data to delete, so I don't think it will be a problem.

Keep looking at its source, will update later...

**Comment by jodydonetti** [comment]
I've finished looking at the code: at first it looks nicely done.

A couple of notes:
- they support both the sync and async api natively, meaning they are not doing async-over-sync or viceversa, but a real double impl (like FusionCache does) which is very nice and will use the underlying system resources best
- they support a pooling mechanism ([here](https://github.com/neosmart/AspSqliteCache/blob/master/SqliteCache/DbCommandPool.cs)) which means the memory allocation will be lower since they reuse existing objects instead of creating new ones every time and consequently, because of that, less cpu usage in the long run because less pressure on the GC (Garbage Collector)
- I also see that they support `CancellationToken`s (by passing it from the public api methods to the underlying db commands and stuff), meaning that it will gracefully handle cancellations in case you need to, like in a mobile app I would imagine that being the app pause/shutdown events or similar

All in all I would suggest you to play with it a little bit, maybe try different with different `CleanupInterval` values + some loadtesting (a lot of writes/reads) to see how it works, but I would say yours is a good approach and that library seems like a good fit 👍

Let me know if you need something more, and if you like how it went!

**Comment by jodydonetti** [comment]
Oh, one last thing: since you talked about cleaning up expired data in SQLite, I'd like to point something out about FusionCache internal behaviour which may be related, so you don't have surprises down the road.

If you'll use the [fail-safe mechanism](https://github.com/jodydonetti/ZiggyCreatures.FusionCache/blob/main/docs/FailSafe.md) in FusionCache keep in mind that the way it works is that the expiration you'll specify will be treated as a "logical expiration" (meaning you app will not see expired data anyway by way of checks done automatically) but the data itself will remain in the cache (so in this case also in SQLite) for more time, specifically based on the `FailSafeMaxDuration`.

Practical example: if you cache something with a `Duration` of `5 min` but enable fail-safe and specify a `FailSafeMaxDuration` of `1 hour`, the data will logically expire after `5 min` but it will be kept in the cache (so in this case in SQLite) for `1 hour`, so that it may be used again in case of problems while updating the the cache in the future. So the internal cleanup in SQLite for expired data will begin not after `5 min` but after `1 hour`.

Hope this helps.

**Comment by seyfeb** [comment]
Hey @jodydonetti , thanks a bunch! I did not begin to expect such an extensive and fast answer 😮 

> So by default every 30 min it will physically remove the data from the db, but if you care about keeping the file size as small as possible you may drop that value to something like 5 min or even less.

That's not what I was looking for. I was only concerned that over a long time (days), the disk cache might fill up the entire disk, if there is no cleanup mechanism involved - so every 30 minutes  or even less is just fine and there are no concerns regarding the performance due to CPU cycles used for cleanup operations. 

Regarding the fail-safe mechanism, thanks for pointing out that objects need to be stored longer than the (logical) expiration time to be able to retrieve them in case of a failure. Although I must admit that this was obvious even to me ;)

The SQLite implementation of the `IDistributedCache` you found ([NeoSmart.Caching.Sqlite](https://www.nuget.org/packages/NeoSmart.Caching.Sqlite/)) looks indeed very promising. It's incredible that you even took the time to have a deeper look into the codebase! I'm definitely going to try this out.

Finally, without having played around with your code, or even used it productively, it looks like a great and well thought-out library so I'm keen to use it. Keep up the good work and thanks for sharing!!

**Reply by jodydonetti** [reply]
Great, happy to help!

**Comment by jodydonetti** [comment]
Hey @seyfeb , just FYI I've been inspired by your question and expanded the related [docs](https://github.com/jodydonetti/ZiggyCreatures.FusionCache/blob/main/docs/CacheLevels.md) to specifically cover disk cache 😉

Thanks!

**Reply by seyfeb** [reply]
That's a good thought. Even though I seem to be the first to ask about the use case it probably is not too exotic. And since you took the effort to investigate this already it makes sense to make this easier to discover :)

**Reply by jodydonetti** [reply]
Hi @seyfeb , how did it went with FusionCache + Disk Cache? How was your experience?

I recently kept circling back about the mobile app scenario you mentioned here, I felt there's was something worth exploring but didn't know what it was. Then, it clicked! The result is that I'm adding a **new option** that would give people in your scenario (and others) an interesting perf boost 🚀

Here is [the issue](https://github.com/ZiggyCreatures/FusionCache/issues/99), if you have some time I would like you to read it and let me know your thoughts.

Thanks!

**Reply by jodydonetti** [reply]
The new [v0.18.0](https://github.com/ZiggyCreatures/FusionCache/releases/tag/v0.18.0) has been released 🎉

It includes this new feature, too.

**Reply by seyfeb** [reply]
Hi @jodydonetti. I must admit I used the setup Fusion + DiskCache with success in a Mac so far build but not on mobile (which is our actual target platform), yet, since there are issues with the Neosmart IDistributedCache implementation when building for/running on Android. Just today we figured out where the issues are and a colleague of mine opened an issue in the repository. So performance-wise the hands-on experience is still limited.

Regarding the linked issue, your reasoning sounds sensible to me and I didn't see any flaws. I don't know how much of a performance boost this will give, but skipping _any_ unnecessary roundtrips is welcome! :)

**Reply by jodydonetti** [reply]
> Hi @jodydonetti. I must admit I used the setup Fusion + DiskCache with success in a Mac so far build

This is already very nice to hear, thanks for sharing! Even though it's something we kinda take for granted with OSS, it's still super interesting to me to hear how a project of mine is being used 😬

> but not on mobile (which is our actual target platform), yet, since there are issues with the Neosmart IDistributedCache implementation when building for/running on Android. Just today we figured out where the issues are and a colleague of mine opened an issue in the repository. So performance-wise the hands-on experience is still limited.

Got it, I hope they fix the issue soon.

> Regarding the linked issue, your reasoning sounds sensible to me and I didn't see any flaws. I don't know how much of a performance boost this will give, but skipping _any_ unnecessary roundtrips is welcome! :)

Exactly, nothing major to be expected, but still less cpu cycles and less memory consumed.

Thanks!

**Reply by seyfeb** [reply]
> Even though it's something we kinda take for granted with OSS, it's still super interesting to me to hear how a project of mine is being used

I totally understand. Seeing the code used in the wild also helps keeping the motivation up for further improving the software ;)
